[WIP] Alternative architecture support for build tooling

detiber opened the pull request:
Requires: https://github.com/openshift/origin/pull/14847

This is similar to: https://github.com/openshift/origin/pull/13365, but attempts to leverage existing image naming and Dockerfiles where possible.

This is still very much a work in progress as I continue to identify and troubleshoot further test failures.

detiber commented:
@mkumatag ptal, this is similar to your PR for adding ppc64le support. Currently I re-use the existing image naming to avoid issues. To publish the images, you could have an external script that would tag and push the images to a remote repository with the arch specific naming.

Eventually we'll want to move to manifest list images, but there is still more plumbing work to be done before we can get to that point.

stevekuznetsov commented on Makefile:
```diff
@@ -270,14 +272,25 @@ install-travis:
 # Example:
 #   make build-rpms
 build-rpms:
+ifeq ($(ARCH),x86_64)
```
Gut feeling is we should push this into the scripts and keep `Makefile` short

detiber replied:
@stevekuznetsov no objections there, but would still need a way to pass in that we only want to build the current architecture? maybe a `OS_ONLY_BUILD_CURRENT_PLATFORM` var? Or maybe treat redistributable as the exception wtih `OS_BUILD_REDISTRIBUTABLE`?

stevekuznetsov commented on hack/build-base-images.sh:
```diff
@@ -5,13 +5,50 @@
+for go_ver in ${golang_versions}; do
+  image_dir="${OS_ROOT}/images/release/golang-${go_ver}"
+  OS_BUILD_IMAGE_ARGS="-f ${image_dir}/Dockerfile${dockerfile_append} --from ${tag_prefix}-base ${ORIG_BUILD_IMAGE_ARGS}"
+  os::build::image "${tag_prefix}-release:golang-${go_ver}" "${image_dir}"
```
need to confirm with @smarterclayton that this makes sense

stevekuznetsov commented on hack/test-go.sh:
```diff
@@ -159,11 +161,25 @@ else
+filtered_test_packages=
+excluded_test_packages=
+for package in ${test_packages}; do
+    if [[ ",${exclude_packages}," =~ ",${package}," ]]; then
```
Not a fan of hack-ey set-contains checks like this, seen them fail too often. Can we set `$IFS` and iterate over the `$exclude_packages` list and do a direct comparison of elements?

stevekuznetsov replied:
Not a fan of hack-ey set-contains checks like this, seen them fail too often. Can we set `$IFS` and iterate over the `$exclude_packages` list and do a direct comparison of elements?

detiber replied:
@stevekuznetsov alternatively, I could just make $exclude_packages space separated, then no need for IFS trickery.

stevekuznetsov commented on images/release/golang-1.8/Dockerfile.altarch:
```diff
@@ -6,23 +6,18 @@
     rpm -V $INSTALL_PKGS && \
     yum clean all && \
-    curl -L https://github.com/google/protobuf/releases/download/v3.0.0-beta-4/protoc-3.0.0-beta-4-linux-x86_64.zip | bsdtar -C /usr/local -xf - && \
-    chmod uga+x /usr/local/bin/protoc && \
```
?

detiber replied:
It's comparing a new Dockerfile for altarch against one from the removed golang-1.6 directory. protobuf-compiler and golang are installed from RPM. The main driver for switching to RPM was that there are no pre-compiled binaries for protoc for ppc64le, aarch64, or s390x.

Similarly, golang doesn't provide pre-compiled binaries for aarch64 or s390x.

stevekuznetsov commented on images/source/cbs-oranges7-test64-common-el7-build.repo:
```diff
@@ -0,0 +1,8 @@
+gpgcheck = 0
+sslverify = 0
+skip_if_unavailable = 1
+includepkgs = golang* protobuf*
```
What `golang` are we bringing in here?

detiber replied:
This repo has builds for ppc64le and aarch64, it is currently a stopgap. These packages will be moved into a repo owned by the CentOS PaaS SIG once we have ppc64le enabled.

In this case, it is providing golang-1.8.3, but only for aarch64 and ppc64le

stevekuznetsov replied:
@sdodson @tdawson may be able to get these into the appropriate 3.6 repos for the PaaS SIG soon?

detiber replied:
@stevekuznetsov I'm working with @tdawson on it, the alt arches will be enabled on the -future repo first. We won't push to the official repos until we've gained some more confidence.

stevekuznetsov commented on images/source/cbs-virt-ovirt-42-el7-build.repo:
```diff
@@ -0,0 +1,7 @@
+enabled = 1
+gpgcheck = 0
+sslverify = 0
+includepkgs = openvswitch*
```
this is for powerpc ovs?

detiber replied:
Yes, as a stopgap.

stevekuznetsov commented on test/util/server/server.go:
```diff
@@ -386,7 +386,7 @@ func IsServerHealthy(url url.URL) (bool, error) {
 	})
 
-	url.Path = "/healthz"
+	url.Path = "/healthz/ready"
```
function is called `IsServerHealthy` not `IsServerReady` so this seems suspect

detiber replied:
@stevekuznetsov that is true, but I'm seeing issues that look to be race condition related because we are trying to test before the api server/controllers are actually ready. Testing against healthz/ready reduced the frequency of the occurrences, but I'm still trying to find a way to verify that the controllers have initialized before letting at least some of the tests go wild.

In particular, I'm seeing a failure with TestOadmPodNetwork and a large chunk of the Router tests.

stevekuznetsov replied:
Just needs to be evaluated on it's own as a change to semantics of this method (and maybe rename it), didn't want it to slip past as a side-effect of this pr

detiber replied:
@stevekuznetsov indeed, I am planning on breaking any integration test fixes into their own PR.

mkumatag commented:
I tried building base images on ppc64le platform and able to build them. :+1:

detiber commented:
@mkumatag I'm working on updating the golang build in CBS to include this patch: https://go-review.googlesource.com/c/45756/ Until then, I suspect you will see intermittent crashes with the builds produced from this process.

mkumatag commented on images/source/Dockerfile:
```diff
@@ -9,5 +9,4 @@ FROM centos:7
 
 ADD origin-local-release.repo /etc/yum.repos.d/
```
You can use `COPY *.repo /etc/yum.repos.d/` to combine all ADD statements.

detiber replied:
Thanks, updated

detiber commented:
@mkumatag The golang build is now ready, re-running `hack/build-base-images.sh` will pull in the golang build with the appropriate fix.

detiber commented:
@stevekuznetsov I still need to pull the test changes out and there is still plenty of room for cleanup, but ptal.

I've added support for cross building the source, base, and release docker images so that the normal jobs can generate those for publishing to docker hub.

I've also added support for running hack/env using a qemu usermode emulated environment, so that alt arch builds/tests can be run on x86_64 systems as well.

detiber commented:
@mkumatag I've updated this PR with some significant changes, can you please take a look?

detiber commented:
rebased for latest changes in master, I've created a separate PR for tracking backporting changes to the release-3.6 branch here: https://github.com/openshift/origin/pull/15323
